platform/linux-generic/odp_pool.c

```diff
@@ -348,6 +348,10 @@ static odp_pool_t pool_create(const char *name, odp_pool_param_t *params,
 		break;
 
 	case ODP_POOL_PACKET:
+		if (params->pkt.min_headroom > CONFIG_PACKET_HEADROOM) {
+			ODP_ERR("Packet headroom size not supported.");
+			return ODP_POOL_INVALID;
+		}
 		headroom    = CONFIG_PACKET_HEADROOM;
```

Comment:
To change this to:

```
headroom = params->pkt.min_headroom;
```

To reflect the API change being proposed in this PR all that's needed is a one line update to the `packet_init()` routine in `odp_packet_internal.h`:

Change 

```
	pkt_hdr->headroom  = CONFIG_PACKET_HEADROOM;
```
to 

```
	pkt_hdr->headroom  = pkt_hdr->buf_hdr.pool_ptr->headroom;
```

However since `pool_ptr` is currently defined as `void *` a few more simple infrastructure changes are needed to undo this "shortcut".

1. `pool_ptr` should be of type `pool_t *`, not `void *` in `odp_buffer_internal.h`

2. `odp_buffer_internal.h` needs to `#include <odp_pool_internal.h>` to get the definition of `pool_t`

3.  Since there is now a circular dependency between `odp_buffer_internal.h` and `odp_pool_internal.h` the `odp_forward_typedefs_internal.h` file has to have the added line: `typedef struct pool_t pool_t;` added to it.

With these changes pools will properly use the headroom requested in by the new `min_headroom` field, as intended.

However, as @psavol is in the process or reworking the packet internals again I'll leave it to him as to whether he'd rather see this change before or after he makes his other changes. An alternate approach would be for pools to define an internal inline API that returns the configured headroom for this pool so `packet_init()` doesn't have to rummage for it.

My point remains is that if we're creating a reference implementation that reference implementation should reflect how we intend the APIs to be used. If we're now going to allow applications to specify headroom size on `odp_pool_create()` then odp-linux should honor those requests.

Reply:
Bala's implementation is correct. Currently, it's much more simpler to always use the default headroom. I'd rather optimize odp-linux for simplicity and performance than for saving memory. When memory usage is a worry, user can rebuild with smaller CONFIG_PACKET_HEADROOM. Also due to data align / cacheline round ups there will be always some headroom - so the potential memory saving in question would be 1 or 2 cachelines per packet, which is not much.